girder/girder_large_image/models/image_item.py

```diff
@@ -39,6 +39,9 @@
 
 
 class ImageItem(Item):
+    _cacheSaveDataLock = threading.RLock()
```

Comment:
@manthey can we use Girder's cache abstraction here (dogpile) so that this plugin is scalable and portable to different execution environments? The use of threading makes me nervous about portability.

Reply:
We are trying to add an associated file to the database, so that a subsequent query doesn't recompute something.  This should be persistent (e.g., a histogram isn't going change once computed), but, since this a cache, a failure isn't particularly critical (it is functionally a cache miss).  These artifacts could be big (e.g., a tile frames object could be an 8k x 8k jpeg).  In the field on an s3 assetstore, the slow point was often storage, so we wanted to return the result when it was ready but lazily cache it so if another requests asks for it, we don't recompute it.

I don't actually know what dogpile is doing or how (or if) it persists data between process restarts.

Reply:
I'd suggest using a cache abstraction that will scale (e.g. memchaced or redis), and preventing stampedes using a key in that cache indicating that the computation is in progress and doesn't need to be repeated. That way we can ensure this code is useful even in multi-process / multi-node deployments.